### ISSUE TITLE
RyuJIT x86: NYI: "Can't decompose expression tree TYP_LONG node"

### ISSUE THREAD

**Issue by pgavlin** [dotnet/coreclr#6982]
This is blocking a number of CQ tests.

**Comment by mikedn** [comment]
I think this is the ADD/SUB NYI that I'm trying to remove in #5884.

**Comment by pgavlin** [comment]
Yes, it is. It looks like the changes you have are essentially good-to-go, perhaps modulo the workaround for #5699.

**Comment by mikedn** [comment]
Yes, the changes should be fine but I hit an assert in LSRA. Something like
```
Assert failure(PID 4752 [0x00001290], Thread: 4140 [0x102c]): Assertion failed 'regRecord->assignedInterval == nullptr' in 'testout1:Func_0_3_1_3_6():long' (IL size 196)
File: d:\j\workspace\x86_ryujit_ch---7ecedcb1\src\jit\lsra.cpp Line: 11258
```
Did you see something like this before?

As for #5699, I think we should just force the x86 JIT to always use an EBP frame to get things going. We can revisit that issue later, it's a performance optimization after all.

**Comment by pgavlin** [comment]
> Did you see something like this before?

I remember having seen this assert before, but the context in which I saw it escapes me. I'll try applying your changes and digging in a bit.

**Comment by mikedn** [comment]
Here's a smallish repro:
```C#
using System;
using System.Runtime.CompilerServices;
class Program {
    [MethodImpl(MethodImplOptions.NoInlining)]
    static void use(long l) {
    }
    [MethodImpl(MethodImplOptions.NoInlining)]
    static long def(int s) {
        return s;
    }
    static void Main() {
        long x = def(4);
        long y = def(2);
        long z = checked(x + x + x);
        use(x);
    }
}
```
Small changes such as removing one of the additions or making `z` used make the assert go away.

**Comment by pgavlin** [comment]
This is due to an GT_LONG node with no use sneaking its way out of decomp + lowering and into LSRA. `TreeNodeInfoInit` for x86 that fails to notice that this node has no use and does not mark it as consuming its operands, which causes LSRA to allocate incorrectly. The fix is at https://gist.github.com/pgavlin/21e57005a7b67edd3796751066d80e2f if you'd like to apply it to your change.

**Comment by pgavlin** [comment]
I decided to file a PR to fix the GT_LONG issue after noticing that it looks like you might also be hitting it in https://github.com/dotnet/coreclr/pull/4986.

**Comment by pgavlin** [comment]
(...and that PR is https://github.com/dotnet/coreclr/pull/6996 :))

**Comment by mikedn** [comment]
I wrote that `TreeNodeInfoInit` code before the LIR changes and back then `gtNext == nullptr` meant "unused" :). Thanks!

**Comment by mikedn** [comment]
Fixed by #5884 ?

**Comment by pgavlin** [comment]
Yes.
